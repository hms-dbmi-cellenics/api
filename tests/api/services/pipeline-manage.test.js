const AWSMock = require('aws-sdk-mock');
const _ = require('lodash');
const AWS = require('../../../src/utils/aws/requireAWS');
const { getQcPipelineStepNames } = require('../../../src/api/services/pipelines/manage/skeletons');
const { getQcStepsToRun } = require('../../../src/api/services/pipelines/manage/getQcStepsToRun');

const mockStepNames = getQcPipelineStepNames();

jest.mock('../../../src/api/services/pipeline-manage/qc-helpers', () => ({
  getQcStepsToRun: jest.fn(() => mockStepNames),
}));

jest.mock('crypto', () => ({
  ...jest.requireActual('crypto'),
  randomBytes: () => Buffer.from('asdfg'),
}));
jest.mock('../../../src/utils/asyncTimer');

const MockExperimentData = {
  Item: {
    sampleIds: {
      L: [
        {
          S: 'oneSample',
        },
        {
          S: 'otherSample',
        },
      ],
    },
    processingConfig: {
      M: {
        doubletScores: {
          M: {
            enabled: {
              BOOL: true,
            },
            filterSettings: {
              M: {
                oneSetting: {
                  N: 1,
                },
              },
            },
            oneSample: {
              M: {
                filterSettings: {
                  M: {
                    oneSetting: {
                      N: 1,
                    },
                  },
                },
                defaultFilterSettings: {
                  M: {
                    oneSetting: {
                      N: 1,
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
};

const { createQCPipeline } = jest.requireActual('../../../src/api/services/pipeline-manage');
const { createGem2SPipeline } = jest.requireActual('../../../src/api/services/pipeline-manage');

describe('test for pipeline services', () => {
  afterEach(() => {
    AWSMock.restore('EKS');
    AWSMock.restore('StepFunctions');
  });

  const mockCluster = {
    cluster: {
      name: 'biomage-test',
      endpoint: 'https://test-endpoint.me/fgh',
      certificateAuthority: {
        data: 'AAAAAAAAAAA',
      },
    },
  };

  const processingConfigUpdate = [
    {
      name: 'doubletScores',
      body: {
        oneSample: {
          filterSettings: {
            oneSetting: 7,
          },
        },
        otherSample: {
          filterSettings: {
            oneSetting: 15,
          },
        },
      },
    },
  ];

  const taskParams = {
    projectId: 'test-project',
    experimentName: 'valerio-massala',
    organism: null,
    input: { type: '10x' },
    sampleIds: ['3af6b6bb-a1aa-4375-9c2c-c112bada56ca'],
    sampleNames: ['sample-1'],
  };

  it('Create QC pipeline works', async () => {
    const describeClusterSpy = jest.fn((x) => x);
    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      describeClusterSpy(params);
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn(
      (stateMachineObject) => _.omit(stateMachineObject, ['definition', 'image']),
    );

    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    const createActivitySpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'createActivity', (params, callback) => {
      createActivitySpy(params);
      callback(null, { activityArn: 'test-actvitiy' });
    });

    const startExecutionSpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'startExecution', (params, callback) => {
      startExecutionSpy(params);
      callback(null, { executionArn: 'test-machine' });
    });

    const getExperimentDataSpy = jest.fn((x) => x);
    AWSMock.mock('DynamoDB', 'getItem', (params, callback) => {
      if (params.TableName.match('experiments')) {
        getExperimentDataSpy(params);
        callback(null, MockExperimentData);
      }
    });

    await createQCPipeline('testExperimentId', processingConfigUpdate);
    expect(describeClusterSpy).toMatchSnapshot();

    expect(createStateMachineSpy.mock.results).toMatchSnapshot();

    expect(getExperimentDataSpy).toHaveBeenCalled();

    expect(createActivitySpy).toHaveBeenCalled();
    expect(startExecutionSpy).toHaveBeenCalled();
    expect(startExecutionSpy.mock.results).toMatchSnapshot();
  });

  it('Parses QC processingConfig correctly', async () => {
    AWSMock.setSDKInstance(AWS);

    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn(
      // eslint-disable-next-line consistent-return
      (stateMachineObject) => (_.cloneDeepWith(JSON.parse(stateMachineObject.definition), (o) => {
        if (_.isObject(o) && o.image) {
          return {
            ...o,
            image: 'MOCK_IMAGE_PATH',
          };
        }

        if (_.isObject(o) && o.ref) {
          return {
            ...o,
            ref: 'MOCK_REF_PATH',
          };
        }
      })),
    );

    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    const createActivitySpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'createActivity', (params, callback) => {
      createActivitySpy(params);
      callback(null, { activityArn: 'test-actvitiy' });
    });

    AWSMock.mock('StepFunctions', 'startExecution', (params, callback) => {
      callback(null, { executionArn: 'test-machine' });
    });

    const getExperimentDataSpy = jest.fn((x) => x);
    AWSMock.mock('DynamoDB', 'getItem', (params, callback) => {
      if (params.TableName.match('experiments')) {
        getExperimentDataSpy(params);
        callback(null, MockExperimentData);
      }
    });

    await createQCPipeline('testExperimentId', processingConfigUpdate);
    expect(createStateMachineSpy.mock.results).toMatchSnapshot();
  });

  it('QC Pipeline is updated instead of created if an error is thrown.', async () => {
    AWSMock.setSDKInstance(AWS);

    const describeClusterSpy = jest.fn((x) => x);
    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      describeClusterSpy(params);
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn((stateMachineObject) => _.omit(stateMachineObject, 'definition'));
    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback({ code: 'StateMachineAlreadyExists' }, null);
    });

    const updateStateMachineSpy = jest.fn((stateMachineObject) => _.omit(stateMachineObject, 'definition'));
    AWSMock.mock('StepFunctions', 'updateStateMachine', (params, callback) => {
      updateStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    const createActivitySpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'createActivity', (params, callback) => {
      createActivitySpy(params);
      callback(null, { activityArn: 'test-actvitiy' });
    });

    const startExecutionSpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'startExecution', (params, callback) => {
      startExecutionSpy(params);
      callback(null, { executionArn: 'test-execution' });
    });

    createQCPipeline.waitForDefinitionToPropagate = () => true;

    await createQCPipeline('testExperimentId', processingConfigUpdate);

    expect(describeClusterSpy).toMatchSnapshot();
    expect(createStateMachineSpy.mock.results).toMatchSnapshot();

    expect(updateStateMachineSpy).toHaveBeenCalled();
    expect(updateStateMachineSpy.mock.results).toMatchSnapshot();

    expect(createActivitySpy).toHaveBeenCalled();
    expect(startExecutionSpy).toHaveBeenCalled();
    expect(startExecutionSpy.mock.results).toMatchSnapshot();
  });

  it('Create Gem2s pipeline works', async () => {
    AWSMock.setSDKInstance(AWS);

    const describeClusterSpy = jest.fn((x) => x);
    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      describeClusterSpy(params);
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn(
      (stateMachineObject) => _.omit(stateMachineObject, ['definition', 'image']),
    );

    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    const createActivitySpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'createActivity', (params, callback) => {
      createActivitySpy(params);
      callback(null, { activityArn: 'test-actvitiy' });
    });

    const startExecutionSpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'startExecution', (params, callback) => {
      startExecutionSpy(params);
      callback(null, { executionArn: 'test-machine' });
    });

    await createGem2SPipeline('testExperimentId', taskParams);
    expect(describeClusterSpy).toMatchSnapshot();

    expect(createStateMachineSpy.mock.results).toMatchSnapshot();

    expect(createActivitySpy).toHaveBeenCalled();
    expect(startExecutionSpy).toHaveBeenCalled();
    expect(startExecutionSpy.mock.results).toMatchSnapshot();
  });


  it('Gem2s Pipeline is updated instead of created if an error is thrown.', async () => {
    AWSMock.setSDKInstance(AWS);

    const describeClusterSpy = jest.fn((x) => x);
    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      describeClusterSpy(params);
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn((stateMachineObject) => _.omit(stateMachineObject, 'definition'));
    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback({ code: 'StateMachineAlreadyExists' }, null);
    });

    const updateStateMachineSpy = jest.fn((stateMachineObject) => _.omit(stateMachineObject, 'definition'));
    AWSMock.mock('StepFunctions', 'updateStateMachine', (params, callback) => {
      updateStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    const createActivitySpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'createActivity', (params, callback) => {
      createActivitySpy(params);
      callback(null, { activityArn: 'test-actvitiy' });
    });

    const startExecutionSpy = jest.fn((x) => x);
    AWSMock.mock('StepFunctions', 'startExecution', (params, callback) => {
      startExecutionSpy(params);
      callback(null, { executionArn: 'test-execution' });
    });

    createGem2SPipeline.waitForDefinitionToPropagate = () => true;

    await createGem2SPipeline('testExperimentId', taskParams);

    expect(describeClusterSpy).toMatchSnapshot();
    expect(createStateMachineSpy.mock.results).toMatchSnapshot();

    expect(updateStateMachineSpy).toHaveBeenCalled();
    expect(updateStateMachineSpy.mock.results).toMatchSnapshot();

    expect(createActivitySpy).toHaveBeenCalled();
    expect(startExecutionSpy).toHaveBeenCalled();
    expect(startExecutionSpy.mock.results).toMatchSnapshot();
  });
});
