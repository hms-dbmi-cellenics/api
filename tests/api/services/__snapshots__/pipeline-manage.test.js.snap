// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for pipeline services Create Gem2s pipeline works 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "biomage-test",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "name": "biomage-test",
      },
    },
  ],
}
`;

exports[`test for pipeline services Create Gem2s pipeline works 2`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "loggingConfiguration": Object {
        "level": "OFF",
      },
      "name": "biomage-gem2s-test-b2bacc54e31beb2bda7b4bd9b1ec33ec84eae60f",
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "tags": Array [
        Object {
          "key": "experimentId",
          "value": "testExperimentId",
        },
        Object {
          "key": "clusterEnv",
          "value": "test",
        },
        Object {
          "key": "sandboxId",
          "value": "default",
        },
      ],
      "type": "STANDARD",
    },
  },
]
`;

exports[`test for pipeline services Create Gem2s pipeline works 3`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "input": "{\\"retries\\":[\\"retry\\"]}",
      "stateMachineArn": "test-machine",
      "traceHeader": undefined,
    },
  },
]
`;

exports[`test for pipeline services Create QC pipeline works 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "biomage-test",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "name": "biomage-test",
      },
    },
  ],
}
`;

exports[`test for pipeline services Create QC pipeline works 2`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "loggingConfiguration": Object {
        "level": "OFF",
      },
      "name": "biomage-qc-test-b2bacc54e31beb2bda7b4bd9b1ec33ec84eae60f",
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "tags": Array [
        Object {
          "key": "experimentId",
          "value": "testExperimentId",
        },
        Object {
          "key": "clusterEnv",
          "value": "test",
        },
        Object {
          "key": "sandboxId",
          "value": "default",
        },
      ],
      "type": "STANDARD",
    },
  },
]
`;

exports[`test for pipeline services Create QC pipeline works 3`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "input": "{\\"samples\\":[{\\"sampleUuid\\":\\"oneSample\\",\\"index\\":0},{\\"sampleUuid\\":\\"otherSample\\",\\"index\\":1}],\\"retries\\":[\\"retry\\"]}",
      "stateMachineArn": "test-machine",
      "traceHeader": undefined,
    },
  },
]
`;

exports[`test for pipeline services Gem2s Pipeline is updated instead of created if an error is thrown. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "biomage-test",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "name": "biomage-test",
      },
    },
  ],
}
`;

exports[`test for pipeline services Gem2s Pipeline is updated instead of created if an error is thrown. 2`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "loggingConfiguration": Object {
        "level": "OFF",
      },
      "name": "biomage-gem2s-test-b2bacc54e31beb2bda7b4bd9b1ec33ec84eae60f",
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "tags": Array [
        Object {
          "key": "experimentId",
          "value": "testExperimentId",
        },
        Object {
          "key": "clusterEnv",
          "value": "test",
        },
        Object {
          "key": "sandboxId",
          "value": "default",
        },
      ],
      "type": "STANDARD",
    },
  },
]
`;

exports[`test for pipeline services Gem2s Pipeline is updated instead of created if an error is thrown. 3`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "stateMachineArn": "arn:aws:states:eu-west-1:test-account-id:stateMachine:biomage-gem2s-test-b2bacc54e31beb2bda7b4bd9b1ec33ec84eae60f",
    },
  },
]
`;

exports[`test for pipeline services Gem2s Pipeline is updated instead of created if an error is thrown. 4`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "input": "{\\"retries\\":[\\"retry\\"]}",
      "stateMachineArn": "arn:aws:states:eu-west-1:test-account-id:stateMachine:biomage-gem2s-test-b2bacc54e31beb2bda7b4bd9b1ec33ec84eae60f",
      "traceHeader": undefined,
    },
  },
]
`;

exports[`test for pipeline services Parses QC processingConfig correctly 1`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "Comment": "QC Pipeline for clusterEnv 'test'",
      "StartAt": "RequestPod",
      "States": Object {
        "CellSizeDistributionFilterMap": Object {
          "ItemsPath": "$.samples",
          "Iterator": Object {
            "StartAt": "CellSizeDistributionFilter",
            "States": Object {
              "CellSizeDistributionFilter": Object {
                "End": true,
                "Parameters": Object {
                  "config": Object {},
                  "experimentId": "testExperimentId",
                  "processName": "qc",
                  "sampleUuid.$": "$.sampleUuid",
                  "server": "remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local",
                  "taskName": "cellSizeDistribution",
                  "uploadCountMatrix": false,
                },
                "Resource": "arn:aws:states:eu-west-1:test-account-id:activity:pipeline-test-mock-uuid",
                "ResultPath": null,
                "TimeoutSeconds": 10800,
                "Type": "Task",
              },
            },
          },
          "Next": "MitochondrialContentFilterMap",
          "ResultPath": null,
          "Type": "Map",
        },
        "ClassifierFilterMap": Object {
          "ItemsPath": "$.samples",
          "Iterator": Object {
            "StartAt": "ClassifierFilter",
            "States": Object {
              "ClassifierFilter": Object {
                "End": true,
                "Parameters": Object {
                  "config": Object {},
                  "experimentId": "testExperimentId",
                  "processName": "qc",
                  "sampleUuid.$": "$.sampleUuid",
                  "server": "remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local",
                  "taskName": "classifier",
                  "uploadCountMatrix": false,
                },
                "Resource": "arn:aws:states:eu-west-1:test-account-id:activity:pipeline-test-mock-uuid",
                "ResultPath": null,
                "TimeoutSeconds": 10800,
                "Type": "Task",
              },
            },
          },
          "Next": "CellSizeDistributionFilterMap",
          "ResultPath": null,
          "Type": "Map",
        },
        "ConfigureEmbedding": Object {
          "Next": "EndOfPipeline",
          "Parameters": Object {
            "config": Object {},
            "experimentId": "testExperimentId",
            "processName": "qc",
            "sampleUuid": "",
            "server": "remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local",
            "taskName": "configureEmbedding",
            "uploadCountMatrix": false,
          },
          "Resource": "arn:aws:states:eu-west-1:test-account-id:activity:pipeline-test-mock-uuid",
          "ResultPath": null,
          "TimeoutSeconds": 10800,
          "Type": "Task",
        },
        "DataIntegration": Object {
          "Next": "ConfigureEmbedding",
          "Parameters": Object {
            "config": Object {},
            "experimentId": "testExperimentId",
            "processName": "qc",
            "sampleUuid": "",
            "server": "remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local",
            "taskName": "dataIntegration",
            "uploadCountMatrix": true,
          },
          "Resource": "arn:aws:states:eu-west-1:test-account-id:activity:pipeline-test-mock-uuid",
          "ResultPath": null,
          "TimeoutSeconds": 10800,
          "Type": "Task",
        },
        "DoubletScoresFilterMap": Object {
          "ItemsPath": "$.samples",
          "Iterator": Object {
            "StartAt": "DoubletScoresFilter",
            "States": Object {
              "DoubletScoresFilter": Object {
                "End": true,
                "Parameters": Object {
                  "config": Object {
                    "oneSample": Object {
                      "defaultFilterSettings": Object {
                        "oneSetting": 1,
                      },
                      "enabled": true,
                      "filterSettings": Object {
                        "oneSetting": 7,
                      },
                    },
                    "otherSample": Object {
                      "enabled": true,
                      "filterSettings": Object {
                        "oneSetting": 15,
                      },
                    },
                  },
                  "experimentId": "testExperimentId",
                  "processName": "qc",
                  "sampleUuid.$": "$.sampleUuid",
                  "server": "remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local",
                  "taskName": "doubletScores",
                  "uploadCountMatrix": false,
                },
                "Resource": "arn:aws:states:eu-west-1:test-account-id:activity:pipeline-test-mock-uuid",
                "ResultPath": null,
                "TimeoutSeconds": 10800,
                "Type": "Task",
              },
            },
          },
          "Next": "DataIntegration",
          "ResultPath": null,
          "Type": "Map",
        },
        "EndOfPipeline": Object {
          "End": true,
          "Type": "Pass",
        },
        "MitochondrialContentFilterMap": Object {
          "ItemsPath": "$.samples",
          "Iterator": Object {
            "StartAt": "MitochondrialContentFilter",
            "States": Object {
              "MitochondrialContentFilter": Object {
                "End": true,
                "Parameters": Object {
                  "config": Object {},
                  "experimentId": "testExperimentId",
                  "processName": "qc",
                  "sampleUuid.$": "$.sampleUuid",
                  "server": "remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local",
                  "taskName": "mitochondrialContent",
                  "uploadCountMatrix": false,
                },
                "Resource": "arn:aws:states:eu-west-1:test-account-id:activity:pipeline-test-mock-uuid",
                "ResultPath": null,
                "TimeoutSeconds": 10800,
                "Type": "Task",
              },
            },
          },
          "Next": "NumGenesVsNumUmisFilterMap",
          "ResultPath": null,
          "Type": "Map",
        },
        "NumGenesVsNumUmisFilterMap": Object {
          "ItemsPath": "$.samples",
          "Iterator": Object {
            "StartAt": "NumGenesVsNumUmisFilter",
            "States": Object {
              "NumGenesVsNumUmisFilter": Object {
                "End": true,
                "Parameters": Object {
                  "config": Object {},
                  "experimentId": "testExperimentId",
                  "processName": "qc",
                  "sampleUuid.$": "$.sampleUuid",
                  "server": "remoter-server-testExperimentId.pipeline-test-namespace.svc.cluster.local",
                  "taskName": "numGenesVsNumUmis",
                  "uploadCountMatrix": false,
                },
                "Resource": "arn:aws:states:eu-west-1:test-account-id:activity:pipeline-test-mock-uuid",
                "ResultPath": null,
                "TimeoutSeconds": 10800,
                "Type": "Task",
              },
            },
          },
          "Next": "DoubletScoresFilterMap",
          "ResultPath": null,
          "Type": "Map",
        },
        "RequestPod": Object {
          "Comment": "Send a message through SNS so that the API assigns a pod to the pipeline",
          "Next": "WaitForPod",
          "Parameters": Object {
            "Message": "{\\"taskName\\":\\"assignPodToPipeline\\",\\"experimentId\\":\\"testExperimentId\\",\\"input\\":{\\"experimentId\\":\\"testExperimentId\\",\\"sandboxId\\":\\"default\\",\\"activityId\\":\\"pipeline-test-mock-uuid\\",\\"processName\\":\\"qc\\"}}",
            "MessageAttributes": Object {
              "type": Object {
                "DataType": "String",
                "StringValue": "PipelineResponse",
              },
            },
            "TopicArn": "arn:aws:sns:eu-west-1:test-account-id:work-results-test-default",
          },
          "Resource": "arn:aws:states:::sns:publish",
          "ResultPath": null,
          "Type": "Task",
        },
        "WaitForPod": Object {
          "ItemsPath": "$.retries",
          "Iterator": Object {
            "StartAt": "GetAssignedPod",
            "States": Object {
              "GetAssignedPod": Object {
                "Comment": "Retrieves first unassigned and running pipeline pod.",
                "Next": "IsPodAssigned",
                "Parameters": Object {
                  "CertificateAuthority": "AAAAAAAAAAA",
                  "ClusterName": "biomage-test",
                  "Endpoint": "https://test-endpoint.me/fgh",
                  "Method": "GET",
                  "Path": "/api/v1/namespaces/pipeline-test-namespace/pods",
                  "QueryParameters": Object {
                    "labelSelector": Array [
                      "type=pipeline,activityId=pipeline-test-mock-uuid",
                    ],
                  },
                },
                "Resource": "arn:aws:states:::eks:call",
                "Type": "Task",
              },
              "IsPodAssigned": Object {
                "Choices": Array [
                  Object {
                    "IsPresent": false,
                    "Next": "NoPodAssigned",
                    "Variable": "$.ResponseBody.items[0]",
                  },
                ],
                "Comment": "Redirects to an error state if the pipeline pod is not assigned yet.",
                "Default": "ReadyToRun",
                "Type": "Choice",
              },
              "NoPodAssigned": Object {
                "Cause": "No available and running pipeline pods.",
                "Error": "NoPodAssigned",
                "Type": "Fail",
              },
              "ReadyToRun": Object {
                "End": true,
                "Type": "Pass",
              },
            },
          },
          "MaxConcurrency": 1,
          "Next": "ClassifierFilterMap",
          "ResultPath": null,
          "Retry": Array [
            Object {
              "BackoffRate": 1.5,
              "ErrorEquals": Array [
                "NoPodAssigned",
              ],
              "IntervalSeconds": 1,
              "MaxAttempts": 13,
            },
          ],
          "Type": "Map",
        },
      },
    },
  },
]
`;

exports[`test for pipeline services QC Pipeline is updated instead of created if an error is thrown. 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "name": "biomage-test",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "name": "biomage-test",
      },
    },
  ],
}
`;

exports[`test for pipeline services QC Pipeline is updated instead of created if an error is thrown. 2`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "loggingConfiguration": Object {
        "level": "OFF",
      },
      "name": "biomage-qc-test-b2bacc54e31beb2bda7b4bd9b1ec33ec84eae60f",
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "tags": Array [
        Object {
          "key": "experimentId",
          "value": "testExperimentId",
        },
        Object {
          "key": "clusterEnv",
          "value": "test",
        },
        Object {
          "key": "sandboxId",
          "value": "default",
        },
      ],
      "type": "STANDARD",
    },
  },
]
`;

exports[`test for pipeline services QC Pipeline is updated instead of created if an error is thrown. 3`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "roleArn": "arn:aws:iam::test-account-id:role/state-machine-role-test",
      "stateMachineArn": "arn:aws:states:eu-west-1:test-account-id:stateMachine:biomage-qc-test-b2bacc54e31beb2bda7b4bd9b1ec33ec84eae60f",
    },
  },
]
`;

exports[`test for pipeline services QC Pipeline is updated instead of created if an error is thrown. 4`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "input": "{\\"samples\\":[{\\"sampleUuid\\":\\"oneSample\\",\\"index\\":0},{\\"sampleUuid\\":\\"otherSample\\",\\"index\\":1}],\\"retries\\":[\\"retry\\"]}",
      "stateMachineArn": "arn:aws:states:eu-west-1:test-account-id:stateMachine:biomage-qc-test-b2bacc54e31beb2bda7b4bd9b1ec33ec84eae60f",
      "traceHeader": undefined,
    },
  },
]
`;
